com.unity.render-pipelines.core/Runtime/Volume/VolumeParameter.cs

```diff
@@ -1641,14 +1642,239 @@ internal override void Interp(VolumeParameter from, VolumeParameter to, float t)
+                float slope = lastKey.outTangent;
+
+                float endValue = lastKey.value + slope * offsetTime;
+                ret = new Keyframe(endTime, endValue, slope, slope);
+            }
+            else
+            {
+                // only remaining case is that we have a proper index
+                ret = keys[index];
+            }
+            return ret;
+        }
+
+        static private void EvalCurveSegmentAndDeriv(ref float dstValue, ref float dstDeriv, Keyframe lhsKey, Keyframe rhsKey, float currTime)
+        {
+            Assert.IsTrue(lhsKey.time <= rhsKey.time);
+
+            float dx = Mathf.Max(rhsKey.time - lhsKey.time, 0.0001f);
+            float dy = rhsKey.value - lhsKey.value;
+            float length = 1.0f / dx;
+            float lengthSqr = length * length;
+
+            float m1 = lhsKey.outTangent;
+            float m2 = rhsKey.inTangent;
+            float d1 = m1 * dx;
+            float d2 = m2 * dx;
+
+            // Note: The coeffecients are calculated to match what the editor does internally. These coeffeceients expect a
+            // t in the range of [0,dx]. We could change the function to accept a range between [0,1], but then this logic would
+            // be different from internal editor logic which could cause subtle bugs later.
 
-        // TODO: Curve interpolation
+            float c0 = (d1 + d2 - dy - dy) * lengthSqr * length;
+            float c1 = (dy + dy + dy - d1 - d1 - d2) * lengthSqr;
+            float c2 = m1;
+            float c3 = lhsKey.value;
+
+            float t = Mathf.Clamp(currTime - lhsKey.time, 0.0f, dx);
+
+            dstValue = (t * (t * (t * c0 + c1) + c2)) + c3;
+            dstDeriv = (t * (3.0f * t * c0 + 2.0f * c1)) + c2;
+        }
+
+        // lhsIndex and rhsIndex are the indices in the keys array. The lhsIndex/rhsIndex may be -1, in which it creates a synthetic first key
+        // at startTime, or beyond the length of the array, in which case it creates a synthetic key at endTime.
+        static private Keyframe EvalKeyAtTime(Keyframe[] keys, int lhsIndex, int rhsIndex, float startTime, float endTime, float currTime)
+        {
+            Assert.IsTrue(keys.Length >= 1);
+            Assert.IsTrue(startTime <= keys[0].time);
+            Assert.IsTrue(keys[keys.Length - 1].time <= endTime);
+
+            Keyframe lhsKey = FetchKeyFromIndexClamped(keys, lhsIndex, startTime, endTime);
+            Keyframe rhsKey = FetchKeyFromIndexClamped(keys, rhsIndex, startTime, endTime);
+
+            float currValue = 0.0f;
+            float currDeriv = 0.0f;
+            EvalCurveSegmentAndDeriv(ref currValue, ref currDeriv, lhsKey, rhsKey, currTime);
+
+            Keyframe dstKey = new Keyframe(currTime, currValue, currDeriv, currDeriv);
+            return dstKey;
+        }
+
+        public override void Interp(AnimationCurve lhsCurve, AnimationCurve rhsCurve, float t)
```

Comment:
Missing unit test for this and checking 0 GC allocation. We don't had time to fully discuss the GC issue today. But yes pooling can be a solution. (What size?) It can perhaps also be possible to cache per AnimationCurve pairs. So it would only create data the first frame then reuse created object next frame.

Public API should always be fully documented, with /// for intellisens (It will create tech debt for us to make it at release of package if it is not there.)

Reply:
Can we use `com.unity.render-pipelines.core\Runtime\Common\ObjectPools.cs`?

Reply:
Closing up the thread on this one. ObjectPools don't seem practical because we have to call Dispose() when each curve is finished per frame. Some of the volumes are in the constant sized stack of 32, and others come from the editor and need to be GC'd normally. That seemed messy enough that we should instead just reuse the LHS, which will always be from the stack of 32.